src/job_token.rs

```diff
@@ -1,139 +1,163 @@
-            parallelism = amt;
+    impl JobServer {
+        pub(super) unsafe fn from_env() -> Option<Self> {
+            let var = var_os("CARGO_MAKEFLAGS")
+                .or_else(|| var_os("MAKEFLAGS"))
+                .or_else(|| var_os("MFLAGS"))?;
+
+            let inner = sys::JobServerClient::open(var)?;
+
+            Some(Self {
+                inner,
+                global_implicit_token: AtomicBool::new(true),
+            })
+        }
+
+        pub(super) fn try_acquire(&self) -> Result<Option<JobToken>, Error> {
+            if !self.global_implicit_token.swap(false, Relaxed) {
+                // Cold path, no global implicit token, obtain one
+                if self.inner.try_acquire()?.is_none() {
+                    return Ok(None);
+                }
+            }
+            Ok(Some(JobToken()))
+        }
+
+        pub(super) fn release_token_raw(&self) {
+            // All tokens will be put back into the jobserver immediately
+            // and they cannot be cached, since Rust does not call `Drop::drop`
+            // on global variables.
+            if self
+                .global_implicit_token
+                .compare_exchange(false, true, Relaxed, Relaxed)
+                .is_err()
+            {
+                // There's already a global implicit token, so this token must
+                // be released back into jobserver
+                let _ = self.inner.release();
+            }
         }
     }
+}
+
+mod inprocess_jobserver {
+    use super::JobToken;
+
+    use std::{
+        env::var,
+        sync::atomic::{AtomicU32, Ordering::Relaxed},
+    };
+
+    pub(crate) struct JobServer(AtomicU32);
 
-    // If we create our own jobserver then be sure to reserve one token
-    // for ourselves.
-    let client = jobserver::Client::new(parallelism).expect("failed to create jobserver");
-    client.acquire_raw().expect("failed to acquire initial");
-    return client;
+    impl JobServer {
+        pub(super) fn new() -> Self {
+            // Use `NUM_JOBS` if set (it's configured by Cargo) and otherwise
+            // just fall back to a semi-reasonable number.
+            //
+            // Note that we could use `num_cpus` here but it's an extra
+            // dependency that will almost never be used, so
+            // it's generally not too worth it.
+            let mut parallelism = 4;
+            if let Ok(amt) = var("NUM_JOBS") {
+                if let Ok(amt) = amt.parse() {
+                    parallelism = amt;
+                }
+            }
+
+            Self(AtomicU32::new(parallelism))
+        }
+
+        pub(super) fn try_acquire(&self) -> Option<JobToken> {
+            let res = self
+                .0
+                .fetch_update(Relaxed, Relaxed, |tokens| tokens.checked_sub(1));
+
+            res.ok().map(|_| JobToken())
+        }
+
+        pub(super) fn release_token_raw(&self) {
+            self.0.fetch_add(1, Relaxed);
```

Comment:
IIUC the relaxed ops in this patch are only correct because we're single threaded. The overhead of using stronger atomics is way smaller than almost anything else `cc` does, so I'm not sure the savings are worth it. More broadly, it's possible that we may find a reason in the future we need to use a thread in some case in the future, on some platform. I'd like it if changing another part of the code doesn't make this slightly wrong.

Assuming I'm not mistaken there, presumably using acquire/release as implied by the function name is right, but feel free to use stronger if that's incorrect.

Reply:
Done, I've changed them to use acquire/release.